common/service/dynamicconfig/fileBasedClient.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamicconfig
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"gopkg.in/yaml.v2"
+)
+
+var _ Client = (*fileBasedClient)(nil)
+
+type constrainedValue struct {
+	Value       interface{}
+	Constraints map[string]interface{}
+}
+
+// FileBasedClientConfig is the config for the file based dynamic config client.
+// It specifies where the config file is stored and how often the config should be
+// updated by checking the config file again.
+type FileBasedClientConfig struct {
+	Filepath     string
+	PollInterval time.Duration
+}
+
+type fileBasedClient struct {
+	values          atomic.Value
+	lastUpdatedTime time.Time
+	config          *FileBasedClientConfig
+	doneCh          chan struct{}
+	logger          log.Logger
+}
+
+// NewFileBasedClient creates a file based client.
+func NewFileBasedClient(config *FileBasedClientConfig, logger log.Logger, doneCh chan struct{}) (Client, error) {
+	client := &fileBasedClient{
+		config: config,
+		doneCh: doneCh,
+		logger: logger,
+	}
+	if err := client.update(); err != nil {
+		return nil, err
+	}
+	go func() {
+		ticker := time.NewTicker(client.config.PollInterval)
```

Comment:
PollInterval may need a default 5s to be defensive.

Reply:
Added a validate function on start to check both poll interval and whether the config file exists.